solver/edge.go

```diff
@@ -116,10 +117,12 @@ type edgeRequest struct {
 	currentKeys  int
 }
 
-// incrementReferenceCount increases the number of times release needs to be
+// takeOwnership increases the number of times release needs to be
 // called to release the edge. Called on merging edges.
-func (e *edge) incrementReferenceCount() {
-	e.releaserCount++
+func (e *edge) takeOwnership(old *edge) {
+	e.releaserCount += old.releaserCount + 1
+	old.owner = e
+	old.releaseResult()
```

Comment:
Question for my own understanding: in `scheduler`, there's `loop` which calls `dispatch` on edges. Once we've done a merge from `old` edge -> `target` edge, will that mean that `dispatch` is only ever called on `target` edge (or if `target` itself has owners, the "ultimate" owner edge after following the linked list) and never be called on `old` edge anymore?

Just wondering because, if true, that would massively simply trying to understand the overall logic. I was initially concerned that we make this `old.releaseResult()` call but then there's still a ton of methods on `edge` that access `e.result` without checking whether `e.owner != nil`, which felt like an avenue to accidentally using an already released result.

But, if we only ever call dispatch on edges with `e.owner == nil`, I think most of those concerns go away since all those other methods should never be called once `e.owner != nil`.

Reply:
> the "ultimate" owner edge after following the linked list) and never be called on old edge anymore?

Yes, the scheduler only accesses edges via `getEdge()` method, and I have added `owner` check into that method in this PR so that once the owner is set, it is returned instead of the old definition.

Reply:
Cool, thanks, that helps.

I really think we should add one or more unit tests for this case if at all possible. Partially for regression but also partially because the unit tests are pretty helpful for understanding the expected behavior too.

Other than that LGTM

Reply:
I'm not sure how tricky it is to repro this with unit test. You need to synchronize the requests to not only cause the triple edge merge, but you need to be able to trigger the evaluation of the second one only after the triple merge has already happened.

@alexcb Maybe your repro can be turned into an integration test?

Reply:
I can definitely give it a try -- I'll use https://github.com/moby/buildkit/blob/86e25b3ad8c20fc420669949f24bb86c74082b2f/solver/scheduler_test.go#L3040 as inspiration.

Reply:
still no luck triggering this via a test; I've been hacking at this function in a branch based off master (i.e. without this fix)

here's what I have so far: https://github.com/alexcb/buildkit/blob/507cc41e4aa901d229e1fa97d65de998120142ff/solver/scheduler_test.go#L3081-L3163 (with the intention of moving it to a new function if I succeed in reproducing it)

I've added in printfs to the scheduler to trace the order needed to reproduce the problem. Here's an example of the error: https://github.com/alexcb/buildkit/blob/507cc41e4aa901d229e1fa97d65de998120142ff/prints-from-reproduction-program

but unfortunately my test isn't triggering the same code path -- one noticable difference is in the test, I only see `newRequestWithFunc called for ...` once per edge, when I believe I need it to occur twice in order to trigger the bad merge.

I'll keep hacking at the test, but wanted to share an update in case there's any obvious parts my test is skipping.

Reply:
I was able to spend some more time with the scheduler_test.go code -- it's pretty neat.

I created some functions to make it easier to wait on cacheKey() returning, and am able to create some edge merges more easily, but I still haven't been able to reproduce the multi edge merge.

here's the updated attempt: https://github.com/alexcb/buildkit/blob/e6da21940e81a000a56c06a59664bd9ab28ea480/solver/scheduler_test.go#L3123

I'm running out of steam on it though, and am missing something timing related I suspect.

Reply:
I think the problem is that you don't only need to create the merge case, but there also needs to be a vertex that depends on the edge that gets merged, and that edge needs to go to the completed state only after the merge has happened. Merging itself does not trigger the bug. What triggers the bug is a new request coming into the vertex that already had the 3-level merge and should never unpark again.

Reply:
Ok, while digging into https://github.com/moby/buildkit/pull/4347#issuecomment-1889353650, I found something that seems to contradict something here.

> > the "ultimate" owner edge after following the linked list) and never be called on old edge anymore?
> 
> Yes, the scheduler only accesses edges via `getEdge()` method, and I have added `owner` check into that method in this PR so that once the owner is set, it is returned instead of the old definition.

My understanding of this is that `dispatch` should *never* be called with an `*edge` that has an `owner`? However, I do seem to see this in logs (using my little investigation branch from my linked comment https://github.com/moby/buildkit/compare/master...jedevc:buildkit:wip-repro-the-deadlock):

```
>>> sha256:614be6d44883c15260f9ce29e2feb9fc4f83bf8467744434f2e6152a747ab383
    sha256:614be6d44883c15260f9ce29e2feb9fc4f83bf8467744434f2e6152a747ab383 owned by sha256:ccb9d6e0f552230dba1da5f989d813984ddb3e5a44cf674e5df1c0e4524f4463
```

Is my understanding wrong here? Is `owner` on a edge a valid input to `dispatch`? And if it is, we should `panic` if we detect it, since some internal assumption seems to be violated then.